shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm

```diff
@@ -87,16 +145,99 @@ - (CGRect)frame {
-    [reversedItems
-        addObject:self.childrenInHitTestOrder[self.childrenInHitTestOrder.count - 1 - i]];
+    SemanticsObject* child = self.childrenInHitTestOrder[self.childrenInHitTestOrder.count - 1 - i];
+    if ([child isKindOfClass:[FlutterScrollableSemanticsObject class]]) {
+      [reversedItems addObject:((FlutterScrollableSemanticsObject*)child).scrollView];
+    } else {
+      [reversedItems addObject:child];
+    }
   }
   return reversedItems;
 }
 
 - (id<UICoordinateSpace>)coordinateSpace {
-  return self.bridge->view();
+  // A regular SemanticsObject uses the same coordinate space as its parent.
+  return self.parent.coordinateSpace ?: self.bridge->view();
+}
+
+@end
+
+/// Scrollable containers interact with the iOS focus engine using the
+/// `UIFocusItemScrollableContainer` protocol. The said protocol (and other focus-related protocols)
+/// does not provide means to inform the focus system of layout changes. In order for the focus
+/// highlight to update properly as the scroll view scrolls, this implementation incorporates a
+/// UIScrollView into the focus hierarchy to workaround the highlight update problem.
+///
+///  As a result, in the current implementation only scrollable containers and the root node
+///  establish their own `coordinateSpace`s. All other `UIFocusItemContainter`s use the same
+///  `coordinateSpace` as the containing UIScrollView, or the root `FlutterView`, whichever is
+///  closer.
+///
+/// Seealso the `frame` method implementation.
+#pragma mark - Scrolling
+
+@interface FlutterScrollableSemanticsObject (CoordinateSpace)
+@end
+
+@implementation FlutterScrollableSemanticsObject (CoordinateSpace)
+- (id<UICoordinateSpace>)coordinateSpace {
+  // A scrollable SemanticsObject uses the same coordinate space as the scroll view.
+  // This may not work very well in nested scroll views.
+  return self.scrollView;
+}
+@end
+
+@interface FlutterSemanticsScrollView (UIFocusItemScrollableContainer) <
+    UIFocusItemScrollableContainer>
+@end
+
+@implementation FlutterSemanticsScrollView (UIFocusItemScrollableContainer)
+
+#pragma mark - FlutterSemanticsScrollView UIFocusItemScrollableContainer Conformance
+
+- (CGSize)visibleSize {
+  return self.frame.size;
+}
+
+- (void)setContentOffset:(CGPoint)contentOffset {
+  [super setContentOffset:contentOffset];
+  // Do no send flutter::SemanticsAction::kScrollToOffset if it's triggered
+  // by a framework update.
+  if (![self.semanticsObject isAccessibilityBridgeAlive] || !self.isDoingSystemScrolling) {
+    return;
+  }
+
+  // The following code assumes the memory layout of CGPoint is exactly the
+  // same as an array defined as: `double coords[] = { x, y };`. Converts
+  // to a Float64List in dart.
+  // The size of a CGFloat is architecture-dependent and it's typically the word
+  // size. The last iOS with 32-bit support was iOS 10.
+  static_assert(sizeof(CGPoint) == sizeof(CGFloat) * 2);
```

Comment:
should we make it a soft-crash or short circuit if this condition does not hold

Reply:
This is a compile-time check. We typically deliver the engine as a prebuilt binary so it's not going to a concern for most users / devs. This assert is mostly for code readers to see the assumptions we made about CGFloat here.

Reply:
This feels pretty naughty. Even just for the sake of readability, I'd be tempted to create a `double[]`, copy the values in, then encode and ship that. The extra CPU work is negligible and the one line of code will save you five lines of comment.